File: src/output/read_s_q_time.m
function read_s_q_time(filename)

s = xmlread(filename);

mylist = s.getElementsByTagName('SQt');

q_temp = zeros([1 mylist.getLength()]);
t_temp = zeros([1 mylist.getLength()]);
Ss_temp = zeros([1 mylist.getLength()]);
Sd_temp = zeros([1 mylist.getLength()]);

n_q = 0; 
last_q = -1000; % q should never be negative I believe
for i = 0 : mylist.getLength()-1
  q_temp(i+1) = str2num(mylist.item(i).getAttribute('q'));
  t_temp(i+1) = str2num(mylist.item(i).getAttribute('t'));
  Ss_temp(i+1) = str2num(mylist.item(i).getAttribute('S-self'));
  Sd_temp(i+1) = str2num(mylist.item(i).getAttribute('S-diff'));
  if q_temp(i+1) > last_q
    n_q = n_q + 1;
    last_q = q_temp(i+1);
  end
end 

top_element = s.getElementsByTagName('s-q-time');

n_time = length(q_temp) / n_q;

q = q_temp(1:n_q);
t = zeros([1 n_time]);
S_d = zeros([n_q n_time]);
S_s = zeros([n_q n_time]);

time_index = 1;
for j = 1 : n_time
  q_index = 1;
  t(time_index) = t_temp(n_q*(j-1)+1);
  for i = 1 : n_q   
    S_d(q_index, time_index) = Sd_temp(n_q*(j-1)+i);
    S_s(q_index, time_index) = Ss_temp(n_q*(j-1)+i);
    q_index = q_index + 1;
  end
  time_index = time_index + 1;
end

subplot(1,2,1)
surf(q, t, S_d')
xlabel('q [AA\^-1]')
ylabel('t [10\^-13 s]')
zlabel('S\_d (q,t)')
title(char(top_element.item(0).getAttribute('title')))

subplot(1,2,2)
surf(q, t, S_s')
xlabel('q [AA\^-1]')
ylabel('t [10\^-13 s]')
zlabel('S\_s (q,t)')

File: src/output/read_s_q_time.py
"""
Used to plot MDMC generated s(q,t) outputs
"""

import xml.etree.ElementTree as ET
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D


def read_s_q_time(filename):
    """Read and plot MDMC generated s(q,t) outputs."""
    
    tree = ET.parse(filename)
    root = tree.getroot()
    
    elements = root.findall('SQt')
    
    q_temp = np.zeros(len(elements))
    t_temp = np.zeros(len(elements))
    Ss_temp = np.zeros(len(elements))
    Sd_temp = np.zeros(len(elements))
    
    n_q = 0
    last_q = -1000  # q should never be negative
    
    for i, elem in enumerate(elements):
        q_temp[i] = float(elem.get('q'))
        t_temp[i] = float(elem.get('t'))
        Ss_temp[i] = float(elem.get('S-self'))
        Sd_temp[i] = float(elem.get('S-diff'))
        if q_temp[i] > last_q:
            n_q += 1
            last_q = q_temp[i]
    
    top_element = root.find('s-q-time')
    
    n_time = len(q_temp) // n_q
    
    q = q_temp[:n_q]
    t = np.zeros(n_time)
    S_d = np.zeros((n_q, n_time))
    S_s = np.zeros((n_q, n_time))
    
    time_index = 0
    for j in range(n_time):
        q_index = 0
        t[time_index] = t_temp[n_q * j]
        for i in range(n_q):
            S_d[q_index, time_index] = Sd_temp[n_q * j + i]
            S_s[q_index, time_index] = Ss_temp[n_q * j + i]
            q_index += 1
        time_index += 1
    
    # Create plots
    fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12, 5), subplot_kw={"projection": "3d"})
    
    Q, T = np.meshgrid(q, t)
    
    # First subplot
    ax1.plot_surface(Q, T, S_d.T)
    ax1.set_xlabel('q [AA^-1]')
    ax1.set_ylabel('t [10^-13 s]')
    ax1.set_zlabel('S_d(q,t)')
    if top_element is not None and top_element.get('title') is not None:
        ax1.set_title(top_element.get('title'))
    
    # Second subplot
    ax2.plot_surface(Q, T, S_s.T)
    ax2.set_xlabel('q [AA^-1]')
    ax2.set_ylabel('t [10^-13 s]')
    ax2.set_zlabel('S_s(q,t)')
    
    plt.tight_layout()
    plt.show()
    
    s_q_time_info = {
        'S_s': S_s,
        'S_d': S_d,
        'q': q,
        't': t
    }
    
    return s_q_time_info


if __name__ == '__main__':
    import sys
    if len(sys.argv) > 1:
        read_s_q_time(sys.argv[1])
